File: ID10267_GraphicalEditor.cpp
/* @JUDGE_ID: 10267 C++ ""Graphical Editor"" */

/* @BEGIN_OF_SOURCE_CODE*/
#include <stdio.h>
#include<iostream>
#include <stdlib.h> /*Character handling functions*/
#include<string.h>


using namespace std;

const int INF = 0x3F3F3F3F;
const int NULO = -1;
const double EPS = 1e-10;

#define MAX 252
char IMG[MAX][MAX];
int M,N;

void cleanIMG(){
	for (int i = 1; i <= N; ++i)
	{
		for (int j = 1; j <= M; ++j)
		{
			IMG[i][j] = 'O';
			IMG[N+1][j] = '0';
		}
		IMG[i][M+1] = '0';
	}
}

void printIMG(){
	for (int i = 1; i <= N; ++i)
	{
		for (int j = 1; j <= M; ++j)
		{
			cout<<IMG[i][j];
		}
		cout<<endl;
	}	
}

void fillRegionR(int x,int y,char c){
	char oldcolor;
	oldcolor = IMG[y][x];

	if (oldcolor == c)
		return;

	IMG[y][x] = c;
	if (x>=1 and IMG[y][x-1]==oldcolor) fillRegionR(x-1,y,c);
	if (x<=M and IMG[y][x+1]==oldcolor) fillRegionR(x+1,y,c);
	if (y>=1 and IMG[y-1][x]==oldcolor) fillRegionR(x,y-1,c);
	if (y<=N and IMG[y+1][x]==oldcolor) fillRegionR(x,y+1,c);

	return;
}

int main() {
	char cmd, c;
	int x,y,x1,x2,y1,y2,tmp;
	char name[128];

	for (int i = 0; i < MAX; ++i)
	{
		for (int j = 0; j < MAX; ++j)
		{
			IMG[i][j] = '0';
		}
	}

	while (true){
		//ler comando entrada
		cin>>cmd;

		switch(cmd){
			case 'I':
				cin>>M>>N;
				//cout<<cmd<<" "<<x<<" "<<y<<endl;
				cleanIMG();
				break;
			case 'C':
				//cout<<cmd<<endl;
				cleanIMG();
				break;
			case 'L':
				cin>>x>>y>>c;
				//cout<<cmd<<" "<<x1<<" "<<y1<<" "<<c<<endl;
				IMG[y][x] = c;
				break;
			case 'V':
				cin>>x>>y1>>y2>>c;
				if( y1> y2){
					tmp = y1;
					y1 = y2;
					y2 = tmp;
				}
				for (int i = y1; i <= y2; ++i)
				{
					IMG[i][x] = c;
				}
				//cout<<cmd<<" "<<x1<<" "<<x2<<" "<<y1<<" "<<c<<endl;

				break;

			case 'H':
				cin>>x1>>x2>>y>>c;
				if (x1 > x2){
					tmp = x1;
					x1 = x2;
					x2 = tmp;
				}
				for (int j = x1; j <= x2; ++j)
				{
					IMG[y][j] = c;
				}
				//cout<<cmd<<" "<<x1<<" "<<x2<<" "<<y1<<" "<<c<<endl;

				break;
			case 'K':
				cin>>x1>>y1>>x2>>y2>>c;
				if (x1 > x2){
					tmp = x1;
					x1 = x2;
					x2 = tmp;
				}
				if( y1> y2){
					tmp = y1;
					y1 = y2;
					y2 = tmp;
				}
				//cout<<cmd<<" "<<x1<<" "<<y1<<" "<<x2<<" "<<y2<<" "<<c<<endl;
				for (int i = y1; i <= y2; ++i)
				{
					for (int j = x1; j <= x2; ++j)
					{
						IMG[i][j] = c;
					}
				}
				break;
			case 'F':
				cin>>x>>y>>c;
				fillRegionR(x,y,c);
				//cout<<cmd<<" "<<x1<<" "<<y1<<" "<<c<<endl;
				break;
			case 'S':
				cin>>name;
				cout<<name<<endl;
				printIMG();
				//cout<<cmd<<" "<<name<<endl;
				break;
			case 'X':
				goto stop;
				break;
			default:
				continue;
		}
		
	}
 	stop:
  return 0;
}

/*@END_OF_SOURCE_CODE*/

File: ID10267_GraphicalEditor.py
# @JUDGE_ID: 10267 Python "Graphical Editor" 

# @BEGIN_OF_SOURCE_CODE


from sys import stdin

def setClear(IMG,M,N):

	for i in range(1,N+1):
		IMG[i][M+1] = '0'
		for j in range(1,M+1):
			IMG[i][j] = 'O'
			IMG[N+1][j] = '0'
	IMG[N+1][M+1] = '0'


def printIMG(IMG,M,N):
	for i in range(1,N+1):
		for j in range(1,M+1):
			print('%c'%(IMG[i][j]),end='')
		print()

def setRegionR(IMG,M,N,x,y, ncolor):

	color = IMG[y][x]
	if color == ncolor:
		return

	p = (x,y)
	q = [p]
	IMG[y][x] = ncolor
	while len(q)>0 :
		xx,yy = q.pop()
		IMG[yy][xx] = ncolor
		if x >= 1 and IMG[yy][xx-1] == color:
			q.append((xx-1,yy))

		if x <=M and IMG[yy][xx+1] == color:
			q.append((xx+1,yy))

		if y>=1 and IMG[yy-1][xx] == color:
			q.append((xx,yy-1))

		if y<=N and IMG[yy+1][xx] == color:
			q.append((xx,yy+1))

	return

if __name__ == '__main__':
	IMG = [['0' for x in range(252)] for y in range(252)]
	M = 0
	N = 0
	while(True):
		line = stdin.readline()
		ll = line.split()
		if ll[0]=='X':
			break

		if ll[0]=='I':
			M = int(ll[1])
			N = int(ll[2])
			setClear(IMG,M,N)

		if ll[0]=='C':
			setClear(IMG,M,N)

		if ll[0]=='L':
			IMG[int(ll[2])][int(ll[1])] = ll[3]

		if ll[0]=='V':
			y1 = int(ll[2])
			y2 = int(ll[3])
			if y1 > y2:
				tmp = y1
				y1 = y2
				y2 = tmp
			for i in range(y1,y2+1):
				IMG[i][int(ll[1])] = ll[4]
			
		if ll[0]=='H':
			x1 = int(ll[1])
			x2 = int(ll[2])
			if x1 > x2:
				tmp = x1
				x1 = x2
				x2 = tmp

			for j in range(x1,x2+1):
				IMG[int(ll[3])][j] = ll[4]
		if ll[0]=='K':
			x1 = int(ll[1])
			x2 = int(ll[3])
			y1 = int(ll[2])
			y2 = int(ll[4])
			if x1 > x2:
				tmp = x1
				x1 = x2
				x2 = tmp
			if y1 > y2:
				tmp = y1
				y1 = y2
				y2 = tmp

			for i in range(y1,y2+1):
				for j in range(x1,x2+1):
					IMG[i][j] = ll[5]

		if ll[0]=='F':
			x = int(ll[1])
			y = int(ll[2])
			xycolor = IMG[y][x]
			ncolor = ll[3]
			#IMG[y][x] = ncolor
			setRegionR(IMG,M,N,x,y,ncolor)

		if ll[0]=='S':
			print(ll[1])
			printIMG(IMG,M,N)



#@END_OF_SOURCE_CODE*/